src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -929,125 +725,316 @@ internal ThreadLocalList(int ownerThreadId)
             /// </summary>
             /// <param name="item">The item to add.</param>
             /// <param name="updateCount">Whether to update the count.</param>
-            internal void Add(T item, bool updateCount)
+            internal void LocalPush(T item, bool needsSync)
             {
-                checked
-                {
-                    _count++;
-                }
-                Node node = new Node(item);
-                if (_head == null)
-                {
-                    Debug.Assert(_tail == null);
-                    _head = node;
-                    _tail = node;
-                    _version++; // changing from empty state to non empty state
-                }
-                else
+                Debug.Assert(Environment.CurrentManagedThreadId == _ownerThreadId);
+                bool lockTaken = false;
+                try
                 {
-                    node._next = _head;
-                    _head._prev = node;
-                    _head = node;
+                    // Full fence to ensure subsequent reads don't get reordered before this
+                    Interlocked.Exchange(ref _currentOp, (int)ConcurrentBagListOperation.Add);
+                    int tail = _tailIndex;
+
+                    // We're going to increment the tail; if we'll overflow, then we need to reset our counts
+                    if (tail == int.MaxValue)
+                    {
+                        lock (this)
+                        {
+                            Debug.Assert(_tailIndex == int.MaxValue, "No other thread should be changing _tailIndex");
+
+                            // Rather than resetting to zero, we'll just mask off the bits we don't care about.
+                            // This way we don't need to rearrange the items already in the queue; they'll be found
+                            // correctly exactly where they are.  One subtlety here is that we need to make sure that
+                            // if head is currently < tail, it remains that way.  This happens to just fall out from
+                            // the bit-masking, because we only do this if tail == int.MaxValue, meaning that all
+                            // bits are set, so all of the bits we're keeping will also be set.  Thus it's impossible
+                            // for the head to end up > than the tail, since you can't set any more bits than all of them.
+                            _headIndex = _headIndex & _mask;
+                            _tailIndex = tail = _tailIndex & _mask;
+                            Debug.Assert(_headIndex <= _tailIndex);
+                        }
+                    }
+
+                    // When there are at least 2 elements' worth of space (and there aren't pending operations that
+                    // would force us to synchronize), we can take the fast path without locking.
+                    if (!needsSync && tail < (_headIndex + _mask))
```

Comment:
I had the same question, actually.  This is copied from ThreadPool:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/ThreadPool.cs#L187-L188
but I think it could probably be changed there as well.  No one else besides this thread is going to change _tailIndex, and even if a concurrent operation steals, that can only increase _headIndex (making more room), so having a single slot remaining (i.e. doing "tail <= (_headIndex + _mask)") should be sufficient.